app/org/sagebionetworks/bridge/validators/SignInValidator.java

```diff
@@ -1,32 +1,53 @@
 package org.sagebionetworks.bridge.validators;
 
+import static org.sagebionetworks.bridge.validators.SignInValidator.RequiredFields.STUDY;
+import static org.sagebionetworks.bridge.validators.SignInValidator.RequiredFields.EMAIL;
+import static org.sagebionetworks.bridge.validators.SignInValidator.RequiredFields.PASSWORD;
+import static org.sagebionetworks.bridge.validators.SignInValidator.RequiredFields.PHONE;
+import static org.sagebionetworks.bridge.validators.SignInValidator.RequiredFields.TOKEN;
+import static org.sagebionetworks.bridge.validators.SignInValidator.RequiredFields.REAUTH;
+
 import static org.apache.commons.lang3.StringUtils.isBlank;
-import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Set;
 
 import org.sagebionetworks.bridge.models.accounts.SignIn;
 
 import org.springframework.validation.Errors;
 import org.springframework.validation.Validator;
 
+import com.google.common.collect.Sets;
+
 public class SignInValidator implements Validator {
     
-    public static final SignInValidator EMAIL_SIGNIN_REQUEST = new SignInValidator(Type.EMAIL_REQUEST);
-    public static final SignInValidator PASSWORD_SIGNIN = new SignInValidator(Type.PASSWORD);
-    public static final SignInValidator EMAIL_SIGNIN = new SignInValidator(Type.EMAIL);
-    public static final SignInValidator REAUTHENTICATION_REQUEST = new SignInValidator(Type.REAUTH);
+    /** Request to sign in via email. */
+    public static final SignInValidator EMAIL_SIGNIN_REQUEST = new SignInValidator(STUDY, EMAIL);
+    /** Sign in using token sent through email. */    
+    public static final SignInValidator EMAIL_SIGNIN = new SignInValidator(STUDY, EMAIL, TOKEN);
+
+    /** Request to sign in via phone. */
+    public static final SignInValidator PHONE_SIGNIN_REQUEST = new SignInValidator(STUDY, PHONE);
+    /** Sign in using token sent through SMS. */
+    public static final SignInValidator PHONE_SIGNIN = new SignInValidator(STUDY, PHONE, TOKEN);
+
+    /** Sign in using an email and password. */
+    public static final SignInValidator PASSWORD_SIGNIN = new SignInValidator(STUDY, EMAIL, PASSWORD);
+    /** Reauthentication. */
+    public static final SignInValidator REAUTH_SIGNIN = new SignInValidator(STUDY, EMAIL, REAUTH);
     
-    private static enum Type {
-        PASSWORD,
-        EMAIL_REQUEST,
+    static enum RequiredFields {
+        STUDY,
         EMAIL,
+        PASSWORD,
+        PHONE,
+        TOKEN,
         REAUTH
     }
     
-    private Type type;
+    private Set<RequiredFields> requiredFields;
 
-    public SignInValidator(Type type) {
-        checkNotNull(type);
-        this.type = type;
+    private SignInValidator(RequiredFields... fields) {
+        requiredFields = Sets.newHashSet(fields);
```

Comment:
Nitpick: This works fine as is, and probably has virtually no impact on performance, but EnumSets are usually preferred for enums.

The only downside is that EnumSets require you to use the EnumSet.of(E first, E...rest) pattern, which could be a problem if for whatever reason, we need an empty SignInValidator. I'll leave it to your judgement.

Reply:
Changed to directly pass enumset to constructor, to avoid an intermediate list.